File: win-prog/ProgWinCS/Lines Curves and Area Fills/BetterPieChart/BetterPieChart.cs
//---------------------------------------------
// BetterPieChart.cs © 2001 by Charles Petzold
//---------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class BetterPieChart: PieChart
{
     public new static void Main()
     {
          Application.Run(new BetterPieChart());
     }
     public BetterPieChart()
     {
          Text = "Better " + Text;
     }
     protected override void DrawPieSlice(Graphics grfx, Pen pen, 
                                          Rectangle rect,
                                          float fAngle, float fSweep)
     {
          float fSlice = (float)(2 * Math.PI * (fAngle + fSweep / 2) / 360);

          rect.Offset((int)(rect.Width  / 10 * Math.Cos(fSlice)),
                      (int)(rect.Height / 10 * Math.Sin(fSlice)));

          base.DrawPieSlice(grfx, pen, rect, fAngle, fSweep);
     }
}

File: win-prog/ProgWinVB/Lines Curves and Area Fills/BetterPieChart/BetterPieChart.vb
'-----------------------------------------------
' BetterPieChart.vb (c) 2002 by Charles Petzold
'-----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class BetterPieChart
    Inherits PieChart

    Shared Shadows Sub Main()
        Application.Run(New BetterPieChart())
    End Sub

    Sub New()
        Text = "Better " & Text
    End Sub

    Protected Overrides Sub DrawPieSlice(ByVal grfx As Graphics, _
            ByVal pn As Pen, ByVal rect As Rectangle, _
            ByVal fAngle As Single, ByVal fSweep As Single)
        Dim fSlice As Single = CSng((2 * Math.PI * _
                                    (fAngle + fSweep / 2) / 360))

        rect.Offset(CInt(rect.Width / 10 * Math.Cos(fSlice)), _
                    CInt(rect.Height / 10 * Math.Sin(fSlice)))

        MyBase.DrawPieSlice(grfx, pn, rect, fAngle, fSweep)
    End Sub
End Class
